esutil/json_reader.go

```diff
@@ -0,0 +1,48 @@
+package esutil
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+)
+
+// JSONReader is an utility function which encodes v into JSON and returns it as a reader.
+//
+func JSONReader(v interface{}) io.Reader {
+	return &jsonReader{val: v, buf: nil}
+}
+
+// JSONEncoder defines the interface for custom JSON encoders.
+//
+type JSONEncoder interface {
+	EncodeJSON(io.Writer) error
+}
+
+type jsonReader struct {
+	val interface{}
+	buf io.ReadWriter
+}
+
```

Comment:
You probably also want to implement https://golang.org/pkg/io/#WriterTo, which will be used by `io.Copy`. That method would avoid allocating an intermediate buffer.

Reply:
Thanks for the suggestion! That makes a lot of sense, I'll have a look into that.

Reply:
Something like this is what I had in mind:

```diff
diff --git a/esutil/json_reader.go b/esutil/json_reader.go
index 2193515..962deab 100644
--- a/esutil/json_reader.go
+++ b/esutil/json_reader.go
@@ -34,30 +34,42 @@ func (r *jsonReader) Read(p []byte) (int, error) {
 }

 func (r *jsonReader) WriteTo(w io.Writer) (int64, error) {
-       if err := r.initialize(); err != nil {
-               return 0, err
-       }
-       return r.buf.WriteTo(w)
+       cw := countingWriter{Writer: w}
+       err := r.encode(&cw)
+       return int64(cw.n), err
 }

 func (r *jsonReader) initialize() error {
        if r.buf == nil {
                r.buf = new(bytes.Buffer)
-               return r.encode()
+               return r.encode(r.buf)
        }
        return nil
 }

-func (r *jsonReader) encode() error {
+func (r *jsonReader) encode(w io.Writer) error {
        var err error

        if e, ok := r.val.(JSONEncoder); ok {
-               err = e.EncodeJSON(r.buf)
+               err = e.EncodeJSON(w)
                if err != nil {
                        return err
                }
                return nil
        }

-       return json.NewEncoder(r.buf).Encode(r.val)
+       return json.NewEncoder(w).Encode(r.val)
+}
+
+type countingWriter struct {
+       io.Writer
+       n int
+}
+
+func (cw *countingWriter) Write(p []byte) (int, error) {
+       n, err := cw.Writer.Write(p)
+       if n > 0 {
+               cw.n += n
+       }
+       return n, err
 }
```